Implement client-side auto-pickup

Rogier-5 opened the pull request:
These patches add support for client-side auto-pickup.

This implementation does not require any changes to the server, or sever-side cooperation, so it is compatible with older servers as well.

Press 'P' to enable auto-pickup; press again to disable. Press multiple times in short succession to change the auto-pickup range.
The key for printing debug stacks was changed to 'O'.

When noclip is active or when moving fast, auto-pickup will be inactive.

Additions:
- Autopickup privilege (requires #4591, and one additional patch (see branch [autopickup-with-client-config](https://github.com/Rogier-5/minetest/tree/autopickup-with-client-config) to be effective).
- Compatibility feature to allow older (unpatched) servers to disable autopickup on the client as well.

tenplus1 commented:
This looks a very good alternative to auto-pickup mods, and I'm glad it can be enabled/disabled.

nerzhul commented on src/content_cao.cpp:
```diff
@@ -696,6 +696,24 @@ v3f GenericCAO::getPosition()
+v3f GenericCAO::getVelocity()
+{
+	ClientActiveObject *parent = getParent();
+	if (parent)
```
please use a ternary operator for this two functions

nerzhul commented:
Interesting, but how does this work if there is two players in range ?

rubenwardy commented:
> Interesting, but how does this work if there is two players in range ?

I imagine exactly the same as if two players clicked the item at the same time - one will win when it resolves server side

lisacvuk commented:
> Interesting, but how does this work if there is two players in range ?
I just tested it with my brother. It seems to be... random?

SmallJoker commented on src/content_cao.cpp:
```diff
@@ -1814,6 +1832,30 @@ bool GenericCAO::directReportPunch(v3f dir, const ItemStack *punchitem,
+		//|| m_velocity != zero
+		// Acceleration may be zero or non-zero
+		//|| m_acceleration != zero
+		|| m_prop.visual != "wielditem";
```
It would be better to check the object name (if possible) instead of trying to guess if it's an item stack or not.

Rogier-5 replied:
>It would be better to check the object name (if possible) instead of trying to guess if it's an item stack or not.

I wrote this code some time ago, and at that time, I couldn't find any dependable way to determine whether it is an itemstack.  don't recall the name being available. But maybe I overlooked something, or maybe it has become available since. I'll have a look at it again.

> Negative function names are confusing. GenericCAO::isItemStack() would be a much better choice here.

They are. I considered this carefully, and given the semantics, I stand by my choice that the name should be negative. See the comment near the top of the function.

`!isItemStack()` does not mean the same as `isNotItemStack()`. There are *three* possibilties:
- certainly not an itemstack
- certainly an itemstack
- not sure.

rubenwardy replied:
Nope, isNotItemStack sounds like !isItemStack
isItemStackFuzzy would be better

t4im replied:
I'm confused. Why wouldn't we know if something is an ItemStack?
If it's possible from Lua, why not from C?

Rogier-5 replied:
There is an `m_name` member, but it is empty for itemstacks. It is used for the player name, so it would be an instance name, instead of a class name.

Rogier-5 replied:
> I'm confused. Why wouldn't we know if something is an ItemStack?
If it's possible from Lua, why not from C?

That's server-side. The server does not send the necessary information to the client.

t4im replied:
Does this help? https://github.com/minetest/minetest/blob/8a98cbda8417c713e03d1ac0be80e88cd2f08a0c/src/content_cao.cpp#L945

There also seems to be an ItemCAO, but i can't find where it is used.

Rogier-5 replied:
I invstigated a bit, and my impression so far is that it can't be used to distinguish whether some active object represents a dropped itemstack. It seems that the best I can do from there is find the (or _a_) defined item that uses the same texture as the active object. If such an item exists, the active object can still be of another type. A falling node for instance.

> There also seems to be an ItemCAO, but i can't find where it is used.

IIUC, ItemCAO is deprecated. All CAOs are now lua-based.

t4im replied:
Sorry, for having moved my comment out here earlier.

> A falling node for instance.

Right, but none of them named `:__builtin:item` which you could look up in the item definition.

SmallJoker commented on src/environment.cpp:
```diff
@@ -2941,4 +2960,210 @@ ClientEnvEvent ClientEnvironment::getClientEvent()
+	}
+
+	// Check items and pick up.
+	if (manual || ((distanceMoved >= pickup_min_movement || elapsedTime >= pickup_idle_interval || prev_quota < 1) && current_quota >= 1))
```
Line has >80 characters

SmallJoker commented on src/environment.cpp:
```diff
@@ -2941,4 +2960,210 @@ ClientEnvEvent ClientEnvironment::getClientEvent()
+
+		std::sort(reachableObjects.begin(), reachableObjects.end());
+
+		for(u32 i=0; i<reachableObjects.size(); i++) {
```
Spaces missing after `for` and around `=`, `<`.

SmallJoker commented on src/environment.h:
```diff
@@ -560,11 +561,34 @@ struct ClientEnvEvent
 {
+	struct PickupKnownCAO
+	{
+		PickupKnownCAO(void) : id(-1), pos(0,0,0), maybe_itemstack(0), picked_up(false), first_seen_time(0), last_seen_time(0) {}
```
Line has >80 characters

SmallJoker commented on src/game.cpp:
```diff
@@ -3087,6 +3094,41 @@ void Game::toggleFog(float *statustext_time, bool *flag)
+			&& g_settings->getBool("always_fly_fast");
+		client->getEnv().toggleAutoPickup(&pickup_range);
+		statustext = std::wstring(L"Auto pick-up enabled; pick-up range is ")
+			+ utf8_to_wide(ftos(int(2 * pickup_range) / 2.0)) + L" block(s)";
```
Blocks are 16x16x16 nodes. We do not want to confuse the players - using meters `m` would be a better choice IMO.

sfan5 replied:
Alternatively call it `nodes`

Rogier-5 commented:
Updated.

The server now sends the type of object (e.g. mobs:cow) to client as well. Auto-pickup uses it if available.

t4im commented:
New to this code, but doesn't `m_gamedef->idef()` already give you access to the item's definition?
https://github.com/minetest/minetest/blob/8a98cbda8417c713e03d1ac0be80e88cd2f08a0c/src/content_cao.cpp#L945


In a quick test i could print the item-name for example for thrown items on a Server, so everything seems to be send.
```cpp
item.getDefinition(idef).name
```

There also seems to be an ItemCAO, but it looks deprecated.

Rogier-5 commented:
> New to this code, but doesn't m_gamedef->idef() already give you access to the item's definition?

But you need an item name to find the item definiton don't you ? Which I don't have...

And apart from that, this is not an item, but an active object that is hiding an item inside... I need to identify the type of active object, not whatever it is hiding, if anything at all.

... or am I overlooking something ?

t4im commented:
> If such an item exists, the active object can still be of another type. A falling node for instance.

You're right, i didn't think about that. Itemframes, too.

HybridDog commented:
Could you disable picking up if the player holds shift or the object's velocity isn't 0?

>But you need an item name to find the item definiton don't you ? Which I don't have...

You could add a client_pickup boolean field to the entity definition, then the server decides whether the player is allowed to auto-pick the object.

Rogier-5 commented:
> Could you disable picking up if the player holds shift or the object's velocity isn't 0?

Already implemented for non-zero velocity.

What is your use-case for the shift key ?

IMO, the shift key is overloaded too much already, and most people will want things to be picked up while crouching. Besides, if you want to temporarily disable pickup, just press `P`.

>> But you need an item name to find the item definiton don't you ? Which I don't have...

>You could add a client_pickup boolean field to the entity definition, then the server decides whether the player is allowed to auto-pick the object.

That is a different change altogether: allowing the server to influence whether the client can do autopickup. It does not solve the problem at hand of how to determine whether an object is an itemstack. And it doesn't work on older severs, which was the objective of that code. For newer servers, I added a change so that the object type is sent to the client, so there is no problem indentifying the type of object.

Rogier-5 commented:
Added autopickup privilege

HybridDog commented:
>What is your use-case for the shift key ? […] Besides, if you want to temporarily disable pickup, just press P.

Some item pickup mods use sneak to determine if the player doesn't want pickup, you're right, since here's a P button, sneak can be ignored or even get some other use.

>And it doesn't work on older severs, which was the objective of that code.

You added the autopickup privilege, can the client enable autopickup if it's an old server which doesn't have the priv yet?

Rogier-5 commented:
> You added the autopickup privilege, can the client enable autopickup if it's an old server which doesn't have the priv yet?

:-(
Didn't think that through long enough...

IMO, autopickup ought to be a default feature, but some servers might want to disable it...

That would be doable with a negative priv (noautopickup). But then it's not a priv anymore...

I'll have to reconsider this... To be continued.

Rogier-5 commented:
Regarding the autopickup privilege: there seems to be no way to communicate generic settings from the server to the client (?). One could be added, and that may happen some time, but that won't solve the issue for older servers.

I'd like to propose a solution which, granted, is a bit hacky, but is usable for both older and newer servers.
- By default, autopickup is enabled.
- If a server owner wants to restrict autopickup, an item definition is created for `clientconfig:restrict-autopickup`:
```lua
minetest.register_craftitem(":clientconfig:restrict-autopickup", {inventory_image = "unknown_item.png"})
```
- If that exists, the client will require the `autopickup`

rubenwardy commented:
That's mega hacky. How about you use a player flag, similar to how minimap is disabled?

Rogier-5 commented:
@rubenwardy I considered that, but the minimap flag is a HUD flag. This has nothing to do with the HUD, so it's a hack too. *and* it pollutes the hud flags.

Do generic player flags exist ?

t4im commented:
Maybe send a `m_prop` property on a per-entity basis.
That would probably allow easy deactivation via setting on the `__builtin:item` but also gives mods an opportunity to add autopickup to other types of entities (powerups maybe?)

Rogier-5 commented:
@t4im I like that idea. Something like `m_prop.collectible` or so...

That doesn't make it compatible with older servers though.

Rogier-5 commented:
Added collectible property to objects.

If set to 0, then the object will never be picked up. If set to a positive value, then it will. If not set (or to a negative value, which is the default), then the type of object determines what happens. I.e. type `__default:item` will be picked up, other objects won't.

HybridDog commented:
you could add cocumentation intop lua_api.txt

Rogier-5 commented:
Documentation added to lua_api.txt

celeron55 commented on doc/lua_api.txt:
```diff
@@ -3400,6 +3400,12 @@ Definition tables
         nametag = "", -- by default empty, for players their name is shown if empty
         nametag_color = <color>, -- sets color of nametag as ColorSpec
         infotext = "", -- by default empty, text to be shown when pointed at object
+        object_type = "<name with which this type of object was registered with 'register_entity'>",
```
object_type shouldn't be in lua_api.txt if mods can never see or set it.

Rogier-5 commented:
Fixed issue in documentation (@celeron55), made some other small changes, and rebased.

* Added an `autopickup_is_privileged` configuration setting (currently only available client-side)
* Renamed server-side entity that disables autopickup on the client to `clientconfig:autopickup_is_privileged`

WIP label can be removed.

Rogier-5 commented:
PR #4591 allows the server to send configuration settings to the client, which will override the values that the client may have configured. This can be used by the server to force autopickup privilege checking on the client. See my branch [autopickup-with-client-config](https://github.com/Rogier-5/minetest/tree/autopickup-with-client-config) for the implementation.

Megaf commented:
Rebase needed, check conflicts. Otherwise looks good.

Rogier-5 commented:
Rebased to latest minetest tree

Megaf commented:
I want to see this tested and merged please.
@nerzhul @est31 @paramat 
Wait for more feedback from me on this mod later today. Will try on a local copy of my server.

est31 commented:
Why not just check for the autopickup privilege and that's it? If its present, autopickup should be allowed, if its not, it should not be. No setting sending or similar needed.

paramat commented:
I disagree with auto-pickup.

Megaf commented:
@est31 I can't change your comment in any way, totally agree with you.

Rogier-5 commented:
> Why not just check for the autopickup privilege and that's it? If its present, autopickup should be allowed, if its not, it should not be. No setting sending or similar needed.

Having the setting is a convenience: to enable auto-pickup globally without the need to grant privs. But you're right that it is not strictly necessary: in a sense, auto-pickup can be considered similar to interact, which must also be granted to each and every player. And interact doesn't have a global setting either.

> Generally I like the idea of this PR though, just android clients need to support it as well.

I'll look into it.

> Also, make sure that you still can drop items without auto pickup removing them again.

Unfortunately, from the client's point of view there isn't any causal relationship between pressing the drop key, and the appearance of an item in the world a short while later. Maybe that a workaround is possible; I was just thinking of a possible approach while writing this.

rubenwardy commented:
how slow are auto pickup mods? I'd rather not add something to core if it's fast enough already.

jastevenson303 commented:
I think it depends on the number of player.  For every entity (let's say, including node drops) they are searching on_step for objects.in.radius(ObjectRef):is_player() and InvRef:room_for_item()... So a single player picking up a singleitem, no problem.

nerzhul commented:
@rubenwardy autopickup is slow in lua generally due to too many calls and a slow SAO, i have a very fast C++ implementation for items in my old fork

HybridDog commented:
Also, if other mods cause lag or you have a bad connection, the visible impact on the delay to the autopickup mod is annoying.

nerzhul commented:
A solution would be having a real client side prédiction with a dedicated sao and CAO, maybe i can port my work on that and enhance it

Ferk commented:
Isn't CSM suitable for this kind of thing so the server doesn't need to do the scanning for entities?

nerzhul commented:
@Ferk not exactly, due to cheating, but a collaboration between server and client can be nice and fast
if your idea is to drop items like magnet, we can imagine when a player is near a dropped item (not dropped by himself, or with a little delay, i implemented this in my fork), the magnet take the item, and in 1 sec, it's added to inventory. Client receive an order of magnetize this item and do a nice animation to get the item

Ferk commented:
@nerzhul Honestly I wasn't thinking about fancy animations, just allowing to send a punch to the items that are already within punching distance of the player (and if they are not within reach there should be a server check for cheating, whether there's autopickup or not).

Though a magnet animation sounds cool.

Wuzzy2 commented:
> Press multiple times in short succession to change the auto-pickup range.

So this allows players to automatically grab items from far away with ease. This is cheap and there's nothing the server can do to prevent this or to customize the range. All the server can do is all-or-nothing. :-1:

What happens if the server already overwrote item entities to add other features, like water flow?

paramat commented on builtin/settingtypes.txt:
```diff
@@ -70,6 +70,36 @@ fast_move (Fast movement) bool false
+
+#    The maximum number of objects that may be picked up per second.
+#    Set to 0 to disable rate limiting.
+autopickup_max_rate (Max rate of auto-pickup) float 3 0
```
Too many settings i feel, these can be hardcoded to practical values:
autopickup_key_toggle_time
autopickup_idle_interval
autopickup_min_movement
autopickup_max_speed
autopickup_max_rate
So, most do not need to be settings, we need to be restrained with settings.

paramat commented on builtin/settingtypes.txt:
```diff
@@ -70,6 +70,36 @@ fast_move (Fast movement) bool false
+#    items are picked up.
+#    When setting the distance using the 'P' key, it will cycle back
+#    to 1 when this value has been reached.
+autopickup_max_range (Maximum auto-pickup range) int 3 0 5
```
Is this max the max player interact distance? I seem to remember it is 4 not 5.

paramat commented on src/guiKeyChangeMenu.cpp:
```diff
@@ -59,6 +59,7 @@ enum
 	GUI_ID_KEY_INVENTORY_BUTTON,
 	GUI_ID_KEY_DUMP_BUTTON,
 	GUI_ID_KEY_RANGE_BUTTON,
+	GUI_ID_KEY_PICKUP_BUTTON,
```
AUTO_PICKUP_BUTTON to be precise?

paramat commented:
Why a privilege? I can't see why only some players would be allowed to autopickup and some not, instead it seems like a choice the client should make through a client setting, as players on a server are likely to strongly prefer enabled or disabled.
There could also be a global server setting for allowing autopickup.

> Added an `autopickup_is_privileged` configuration setting (currently only available client-side)

The setting name is unclear, do you mean 'autopickup_priv_granted` or 'autopickup_allowed`?

Does this PR animate the items so they move toward the player? I'm not asking for that, i doubt that is needed, but without that maybe an 'autopickup sound' is needed to indicate an autopickup when an item disappears.

Overall i think this needs some changes, reconsideration of how autopickup is enabled and by what (client / server), and there are many unnecessary settings added.
:-1: For how it is now, but it could be ok with changes.

paramat commented:
@Rogier-5 does this need adopting by someone else or will you work on it further?

Wuzzy2 commented:
I agree that implementing this as a privilege does not make sense. It's a gameplay feature. Let's not repeat the same mistake we did with `zoom`.

paramat commented:
Author disappeared, needs adoption and a rethink, close in 1 month if no adoption commitment.

Fixer-007 commented:
Wonder if author got tired of waiting and just moved on past minetest.

rubenwardy commented:
This is perfectly possible by mods already, and adds unneed complexity

paramat commented:
Fixer, to be fair, at 10 Oct 2016 the author was intending to do develop this further and address points, but never did, so it's not really a case of the author waiting for dev attention (which of course happens elsewhere).

paramat commented:
:-1: 
Neglected, low priority and disapproved, closing.
